### PR TITLE
Post items

### PR DESCRIPTION
**Pull request by dtrannam**
This PR creates a button that allows users to save items/collections. The button is self contained so we can move it to another section if needed. For now, we can navigate to it by going to:

>  http://localhost:5173/dashboard.

Most of this PR can be found in. The other file changes appear to be due to prettier.
>  src/components/NewItemForm.jsx

- Error Handling:
  - If we get a 403, this will redirect us to the login-signup page
  - Standard client validation to check for null fields
  - All other error will be generic

- New Button
Hiome

- Clicking New will open a menu
Options

- Saving a collection
saveCollection

- Saving a vault item (will console.log for now)
saveVaultItem

- Collections as a menu item in vault item
saveVaultItemCollections

### CODE REVIEW

File: src/components/NewItemForm.jsx
Comment (angieyeh) [inline src/components/NewItemForm.jsx:58]: For both vault collections and vault items, when a user closes the menu without submitting the form and opens up the model again, the form is still populated with the old values.

File: src/components/NewItemForm.jsx
Comment (angieyeh) [inline src/components/NewItemForm.jsx:178]: The server will return a newly created vault item if the response is a success. In a previous conversation we were saying we should only update the UI when the response is a success. How were you thinking about handling updating the UI when creating new vault items and vault collections?

File: src/components/NewItemForm.jsx
Comment (angieyeh) [inline src/components/NewItemForm.jsx:81]: On vault creation, the server returns the newly created vault. We could update the state with just that one item. Since we don't have to decrypt the vault collections this is likely fine too.

**Review by angieyeh — CHANGES_REQUESTED** [review]
Nice job! I think it's probably faster in the short term to keep all the vault collection and items creation in one file, but it has grown pretty big.

I was looking at the modal titles (not part of this PR) , I think we may want to use different language to describe a vault item. Maybe Add New Account or Add New Item or like bw they just have "NEW ITEM".

File: src/components/NewItemForm.jsx
Comment (dtrannam) [reply src/components/NewItemForm.jsx:58]: Good catch! Updated

File: src/components/NewItemForm.jsx
Comment (dtrannam) [reply src/components/NewItemForm.jsx:81]: Yes good call out, updated

File: src/components/NewItemForm.jsx
Comment (dtrannam) [reply src/components/NewItemForm.jsx:178]: One possible option I could think would work is to extract the POST request onto the parent and pass in the data as props. This way we can access both the new item and the collection in the dashboard (assuming that's where we keep it). I'm not sure if that's the best option, but probably the one that comes to mind. Open to suggestions!

File: src/components/NewItemForm.jsx
Comment (derthadams) [inline src/components/NewItemForm.jsx:1]: I'm not sure why it's happening, but when I try filling out either the "Add a new collection" or "Add a new vault Item" form, tabbing to get to the next field closes the modal. If you then click on the "NEW" button again, it bypasses the "Collections/Vault Items" submenu and displays the form in the state that you left it.


https://github.com/secure-password-manager/PasswordManagerWeb/assets/39425112/ca3161ac-b2c7-4b2a-ba2f-389fa22c06f0

Comment (angieyeh) [reply src/components/NewItemForm.jsx:1]: I also noticed the issue with c behaving strangely like what you showed us in the demo. Maybe it's related to this tabbing issue.

File: src/config/AppConstant.js
Comment (angieyeh) [inline src/config/AppConstant.js:6]: What's the reason we make Collections plural but Item singular?

File: src/config/AppConstant.js
Comment (angieyeh) [inline src/config/AppConstant.js:15]: Nit: I would export these URLs out alphabetically, but it is a personally preference so that I can find things more easily

File: src/config/AppConstant.js
Comment (angieyeh) [reply src/config/AppConstant.js:6]: It would make the URLS long, but I would consider adding VAULT as the first word of the URLs so that we can easily find all the vault related URLs.

File: src/components/NewItemForm.jsx
Comment (dtrannam) [reply src/components/NewItemForm.jsx:1]: Good catch on the tab issue, I fixed it. Seems like it's a well known fix. I'm not seeing much on the "c" error (although I am getting it now. I'll continue looking into it, but will mark it as resolved for now to get this moving

File: src/config/AppConstant.js
Comment (dtrannam) [reply src/config/AppConstant.js:6]: I think I was doing this aimlessly, I will fix this with the routes PR for next week just so it doesn't cause merge conflicts with Derth's next merge here. I'll add a note to fix this and the alphabetically order on export! 

* https://github.com/secure-password-manager/PasswordManagerWeb/pull/53/files

File: src/components/NewItemForm.jsx
Comment (angieyeh) [inline src/components/NewItemForm.jsx:92]: Nice I was just going to ask if there was a way to abstract the try catch somehow to handle the errors since we will be needing the error handler in multiple places!

File: src/components/NewItemForm.jsx
Comment (angieyeh) [reply src/components/NewItemForm.jsx:169]: As we talked about in our FE sync today, we will pass the data down as props to the components that need the data. We will pessimistically update the UI. This will need to be done after the GET and UI Layout for displaying VC and VI are complete.

File: src/components/NewItemForm.jsx
Comment (angieyeh) [inline src/components/NewItemForm.jsx:121]: This function is for retrieving collections right?

File: src/components/NewItemForm.jsx
Comment (angieyeh) [inline src/components/NewItemForm.jsx:128]: Unless I'm missing something we don't encrypt the vault collection  names using they key so I don't think we need it.

File: src/components/NewItemForm.jsx
Comment (angieyeh) [inline src/components/NewItemForm.jsx:84]: In my text editor it says data1 is not defined. Do you see that error in your editor?

File: src/components/NewItemForm.jsx
Comment (angieyeh) [inline src/components/NewItemForm.jsx:90]: My text editor says this is also undefined.

File: src/components/NewItemForm.jsx
Comment (angieyeh) [inline src/components/NewItemForm.jsx:196]: We also need an additional call to set the default collection id. In the vault items form it is still showing the last selected folder when I open the modal again.

**Review by angieyeh — CHANGES_REQUESTED** [review]

File: src/common/ServerAPI.jsx
Comment (angieyeh) [inline src/common/ServerAPI.jsx:57]: Sorry I keep seeing other things to ask about after I leave a review. 

Are we creating a vault item here? Should we use ITEMS_URL instead?

File: src/components/NewItemForm.jsx
Comment (dtrannam) [reply src/components/NewItemForm.jsx:121]: Yes! I used getUserItems for collections and the keys. I combined it into one as I figured we'd either get both or throw an error

File: src/components/NewItemForm.jsx
Comment (dtrannam) [reply src/components/NewItemForm.jsx:84]: Yes! Sorry this was a testing line that should have been removed

File: src/components/NewItemForm.jsx
Comment (dtrannam) [reply src/components/NewItemForm.jsx:128]: The call here is to get the key to be used later with vault items and it won't be used on collections. I decided to combine both the collections and the get user key in one call, but I can see now that the naming can lead to confusion. I will rename the function from getUserItems to maybe getUserDetails

File: src/components/NewItemForm.jsx
Comment (angieyeh) [inline src/components/NewItemForm.jsx:170]: Is there a reason this is still commented out? Now that the symmetric key is in the global state (need to get latest from master), you should be able to access the key and then make a request to create vault items. I think this part is necessary for this PR to be complete.

**Review by angieyeh — CHANGES_REQUESTED** [review]
In addition to fixing merge conflicts, the changes on master should unblock you the parts you've been blocked on!

File: src/components/NewItemForm.jsx
Comment (dtrannam) [reply src/components/NewItemForm.jsx]: I left it as setCollectionID(defaultCollectionsID); for now in the next PR

File: src/components/NewItemForm.jsx
Comment (dtrannam) [reply src/components/NewItemForm.jsx]: This should be updated now

File: src/components/NewItemForm.jsx
Comment (dtrannam) [reply src/components/NewItemForm.jsx]: This should be updated now

File: src/components/NewItemForm.jsx
Comment (derthadams) [reply src/components/NewItemForm.jsx]: @angieyeh It should be possible to prop-drill collections from the DashboardPage component assuming this component is an ancestor.

The symmetric key will be needed in this component for encrypting the user data before posting it to the server. You can [retrieve the symmetric key from the global store directly](https://github.com/secure-password-manager/PasswordManagerWeb/blob/627b7be918356fbfa3230a008d5f874502e996e8/src/routers/DashboardPage.jsx#L15).

File: src/config/AppConstant.js
Comment (angieyeh) [inline src/config/AppConstant.js:17]: NIT: USER_KEY_URL

**Review by angieyeh — APPROVED** [review]
NICE JOB and LGTM!!! 🎉  I like how you handle all the form errors!

**Pull request merged**
